src/ucs/datastruct/bitmap.h

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_BITMAP_H_
+#define UCS_BITMAP_H_
+
+#include <ucs/sys/compiler_def.h>
+
+BEGIN_C_DECLS
+
+#define _UCS_BITMAP_BITS_IN_WORD (64)
+#define _UCS_BITMAP_WORD_INDEX(_bit_index) ((_bit_index) / _UCS_BITMAP_BITS_IN_WORD )
+#define _UCS_BITMAP_BIT_IN_WORD_INDEX(_bit_index) ((_bit_index) % _UCS_BITMAP_BITS_IN_WORD)
+
+/*
+ * Represents an n-bit bitmap, where n is a multiple of 64, by using an array
+ * of 64-bit unsigned long integers. Memory is allocated on stack
+ * 
+ * @param _bits Bits number in the bitmap
+ */
+#define _UCS_BITMAP_DECLARE_TYPE(_bits) \
+    typedef struct { \
+        uint64_t bits[_UCS_BITMAP_WORD_INDEX(_bits)]; \
+    } ucs_bitmap_t(_bits);
+
+/**
+ * Expands to bitmap type definition
+ *
+ * @param _bits Bits number (as passed to UCS_BITMAP_DECLARE_TYPE)
+ *
+ * Example:
+ *
+ * @code{.c}
+ * ucs_bitmap_t(64) my_bitmap;
+ * @endcode
+ */
+#define ucs_bitmap_t(_bits) \
+    _UCS_BITMAP_IDENTIFIER(_bits, _t)
+
+#define _UCS_BITMAP_IDENTIFIER(_name, _suffix) \
+    UCS_PP_TOKENPASTE(ucs_bitmap_, UCS_PP_TOKENPASTE(_name, _suffix))
+
+_UCS_BITMAP_DECLARE_TYPE(64)
+_UCS_BITMAP_DECLARE_TYPE(128)
+
+/**
+ * Get bit value
+ * 
+ * @param _bitmap Bitmap to get from
+ * @param _index Bit index
+ * 
+ * @return Bit value (0 or 1)
+*/
+#define UCS_BITMAP_GET(_bitmap, _index) \
+    (((_bitmap)->bits[_UCS_BITMAP_WORD_INDEX(_index)] >> \
+    _UCS_BITMAP_BIT_IN_WORD_INDEX(_index)) & 1)
+
+/**
+ * Set a bit
+ * 
+ * @param _bitmap Bitmap to set in
+ * @param _index Bit index
+*/
+ #define UCS_BITMAP_SET(_bitmap, _index) \
+    ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] |= \
+    ((uint64_t) 1ull << _UCS_BITMAP_BIT_IN_WORD_INDEX(_index));
+
+/**
+ * Unset a bit
+ * 
+ * @param _bitmap Bitmap to unset in
+ * @param _index Bit index
+*/
+ #define UCS_BITMAP_UNSET(_bitmap, _index) \
+     ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] &= \
+     ~((uint64_t) 1 << _UCS_BITMAP_BIT_IN_WORD_INDEX(_index));
+
+/**
+ * Clear a bitmap by setting all bits to zero
+ * 
+ * @param _bitmap Bitmap to clear
+ * 
+ * @return Value of the bit (0 or 1)
+*/
+#define UCS_BITMAP_CLEAR(_bitmap) \
+    (bzero((_bitmap)->bits, sizeof((_bitmap)->bits)))
+
+/**
+ * Iterate over all set bits of a given bitmap
+ * 
+ * @param _index Bit index (relative to the *whole bitmap*)
+ * @param _work Index of the 64-bit word in which the bit is set
+ * @param _bitmap Bitmap to iterate on
+ */
+#define UCS_FOR_EACH_BIT(_index, _word, _bitmap) \
```

Comment:
1. i'd specify `bitmap` first as in all other macros working with it
2. do we want to expose global bit index somehow?

Reply:
we already have macro with same name in src/ucs/sys/math.h:150

Reply:
Changed the name. The `index` is already exposed (`_index` var).

Reply:
i meant index in the bitmap, not in the word. Also please move `_bitmap` to be the first param

Reply:
what about simplify macro:
```
#define UCS_FOR_EACH_BIT(_index, _word, _bitmap) \
    for (_word = 0; _word < ucs_array_size((_bitmap)->bits); _word++) \
        ucs_for_each_bit (_index, (_bitmap)->bits[_word])
```

Reply:
@brminich  - I wanted to but `_bitmap` first because it's more logical, but I noticed that `ucs_for_each_bit` (in `math.h`) puts `_index` first, so I wanted to be consistent with existing code. What do you say?

Reply:
not sure why it is done this way in `ucs_for_each_bit`, i'd add `_bitmap` first for consistency with other bitmap funcs
I think global index is important, otherwise user will have to calculate it manually.
For example:
```
UCS_FOR_EACH_BIT(index, word, my_bitmap) {
    if (some_condition) {
        // here I want to clear the bit
    }
 }
```

Reply:
> not sure why it is done this way in `ucs_for_each_bit`

C loop: ```for (index = value; condition_to_continue; iterator_next)```

Reply:
@hoopoepg ,
I can't just use ` ucs_for_each_bit` out-of-the-box. We need 2 vars here (`index` & `index_in_word`) and we can't run any code in the iteration body in the macro as it would break the macro syntax

Reply:
reuse code with ucs_for_each_bit

Reply:
and linked list enumerator:
```#define ucs_list_for_each(_elem, _head, _member)```

Reply:
Michael suggested, we can simply iterate over all bits of all words here (unlike `ucs_for_each_bit` which uses more efficient BSFQ).
It will be more clearer and readable code, as well as simpler to use and less error-prone.
But from the performance side it will be obviously worse. What do you think? @yosefe